src/tribler-gui/tribler_gui/widgets/triblertablecontrollers.py

```diff
@@ -287,7 +293,7 @@ def add_menu_item(self, menu, name, item_index, callback):
     def selection_can_be_added_to_channel(self):
         for row in self.table_view.selectionModel().selectedRows():
             data_item = row.model().data_items[row.row()]
-            if 'type' in data_item and data_item['type'] in (REGULAR_TORRENT, CHANNEL_TORRENT, COLLECTION_NODE):
+            if dict_has(data_item, 'type', REGULAR_TORRENT, CHANNEL_TORRENT, COLLECTION_NODE):
```

Comment:
Actually, I don't like this new `dict_has` function. In my opinion, it makes code much harder to understand.
1. Previously the code did not require additional context to understand what is going on. Now I need to look at the `dict_has` source to understand the details of its logic.
2. In the previous code it was clear to me that `dict_item` is a `dict` instance (or, at least, some objects with the Mapping interface). Now the code will not break if `dict_item` is, for example, `None`. So, if I need to extend this function in the future, I need to spend time to answer the question "can dict_item be anything other than dict", whereas in the previous version it was clear.
3. The new version is not significantly shorter. If the goal is to achieve brevity it was possible to write instead:
```python
if data_item.get('type') in (REGULAR_TORRENT, CHANNEL_TORRENT, COLLECTION_NODE):
    ...
```

So I propose to roll back this change and remove the `dict_has` function.

Reply:
Python is a dynamically typed language but it is strongly typed. The expression`foo[bar]` generates `KeyError` exception when it is incorrectly assumed that `foo` should be a mapping and `bar` is a required item of this mapping, and it is a good thing. It is like a programmer explicitly adds some assertions to its code which will guarantee that errors will raise fast and not be propagated to different places. When we add unnecessary checks `if foo` or `if bar in foo` we essentially convert strongly typed language to a weakly typed one.

Reply:
> Previously the code did not require additional context to understand what is going on. Now I need to look at the dict_has source to understand the details of its logic.

Yes, and this is ok. You need to understand once what this function does, and then relax for all occurences.
`dict_has` just extract a pattern that was used in this file three times

> Now the code will not break if data_item is, for example, None
Yes, and this is great. No excuses for crash on the user side.

> The new version is not significantly shorter.
Yes, but shortnesses is doubtful purpose. The new version reduces the number of repetitions (of a pattern and of key's value, inside the pattern). Therefore it decreases the chance to make a mistake.

Reply:
> dict_has just extract a pattern that was used in this file three times

Not all patterns probably should be extracted. `a + b * c` is better than `addmul(a, b, c)`

> Yes, and this is great. No excuses for crash on the user side.

If we a sure that, say, something can never be `None`, adding unnecessary handling for `None` leads to less clear code, as from now it says to the future reader: there are some obscure cases when this variable really can be `None`.

> Yes, and this is ok. You need to understand once what this function does, and then relax for all other occurrences.

Many developers need to do it, and even a single developer does not remember all details of all the small functions that he read in the entire codebase and needs to read it again.

Actually, this specific case is small enough to turn it into the bikeshedding, but the general pattern looks a bit like a code obfuscation to me.

Reply:
I agree with @kozlovsky . Python is not C++: Python programmers are used to some idiomatic ways to do simple stuff like checking `dict` attributes, etc. Python is not about inventing your own sub-language as C++ does. Python is about making the code immediately readable to anyone.


So, replacing an explicit one-liner with an implicit one-liner goes against [the Zen of Python](https://en.wikipedia.org/wiki/Zen_of_Python)

`import this` :wink:

Reply:
> Not all patterns probably should be extracted. a + b * c is better than

Yes, not all. But some -- should.

> there are some obscure cases when this variable really can be None

Yes, and if we take a look at the recent issues from the GUI side, we will be ensured that there are a lot of "obscure cases". So, here the question, what is better: to let the client crash at the next "obscure case" and fix it, or to assume that there are such cases, and carefully handle them?

Reply:
> what is better: to let the client crash at the next "obscure case" and fix it, or to assume that there are such cases, and carefully handle them?

To let the client crash at the next few obscure cases, fix 3-4 of these and then figure out the common pattern and refactor it for good. Alas, the problem with the GUI (especially with the PyQT one) is that there are far too many corner cases to devise a consistent solution upfront...

Reply:
@ichorid client's crash should be unacceptable for a client-oriented product. If we want a million users, we have to take more care of them.

Reply:
Silencing errors results in accumulation of unseen defects. What is worse: a button that crashes the application instantly, or a button that does not do anything?

Reply:
I'll give you an answer from my POV, assuming that we talking about Tribler (not in general):
"a button that does not do anything" is better than "a button that crashes the application instantly"

Reply:
How do you know that your buttons work for your users then?

Reply:
I assume that the author of the button checked the button's correctness when he added the corresponding feature. 
Plus all new features should be covered by tests.

But even if no. It is better to receive a user's report that "the button doesn't work" than receive the Sentry event "application is crashed".

In the first case, we will get a loyal user, in the second case, we will lose him.

Reply:
>But even if no. It is better to receive a user's report that "the button doesn't work" than receive the Sentry event "application is crashed". In the first case, we will get a loyal user, in the second case, we will lose him.

What's the probability of the first case vs the second? How many users write reports about broken buttons? More important, *what kind of user* is going to produce these reports? Is it new users, or experienced/loyal ones? What's the price of a broken button in terms of silently disappointed users?

We [already had](https://github.com/Tribler/tribler/issues/4930) this kind of discussions before. The consensus was: "We are going to make Tribler work in two modes: either *debug* or *production*. Debug should always crash, production should never crash. Production release should only be done after a couple of weeks of RCs in Debug mode." No one ever bothered to implement it properly, though.

Well, we don't do extensive RCs now, since we are on the monthly release schedule. So, every release must **look** as stable as possible, even if it means hiding defects, right?

Reply:
The number of safety checks which can be added to the code "just for a case" is very big. If we start adding such safety checks to every possible place then it would be very hard to understand, which of these safety checks are real checks added to prevent some possible situations. As in the current code we don't have situations when `data_item` can be None or when it might do not contain "type" key, in my opinion, it is better not to include such checks. Otherwise, if we start applying consistently the same approach to all possible places in code, the codebase will be much harder to read.

Reply:
>  As in the current code we don't have situations when data_item can be None or when it might do not contain "type"

How can you prove that?

Reply:
> How can you prove that?

By the absence of the corresponding exceptions in reports

Reply:
Dijkstra would say that exceptions in reports can be used to show the presence of bugs, but never to show their absence :)

Reply:
@ichorid 

> So, every release must look as stable as possible, even if it means hiding defects, right?

Probably it is better to just answer "yes" on your question.

But I would say, if release doesn't look stable, you should publish it. Taking into account, that "obscure cases" always will be presented... we need to make release stable with the margin of safety.

(in my humble view)

Reply:
> But I would say, if release doesn't look stable, you should publish it.

Maybe you meant "you **shouldn't** publish it"? :wink:

Reply:
Yes 🤦 
Fixed.

Reply:
> Dijkstra would say that exceptions in reports can be used to show the presence of bugs, but never to show their absence :)

Of course. Yet exceptions is almost the only kind of feedback and support that we get from our userbase. Because, you know, we can't do telemetry.